_use_cases/schema_validation_and_enforecement.markdown

```diff
@@ -0,0 +1,84 @@
+
+To overcome the problem, the Kafka **client** ecosystem supports _schemas_.  Schemas provide a programmatic description
+of the message format.
+
+Producers use a schema to ensure that they produce messages that adhere to the agreed message format.  The consumer
+uses the same schema to help it decode the incoming message with the certainty that it conforms to the expected format.
+
+The schemas themselves could be packaged as part of the application.  Whilst the approach is functional, it is quite
+inflexible as it means all applications need to be redeployed every time a schema changes.  This becomes a challenge
+as the number of applications in the system grows.  To counter this challenge organizations often choose to deploy a
+*Schema Registry*. 
+
+##### Schema Registry
+
+The role of the  **Schema Registry**  is to maintain a library of schemas that are in-use within an organisation.
+
+Producer applications call the registry to retrieve the schema for the messages they need to send.
+
+Consuming applications use information embedded in the incoming message to be able to retrieve the right schema from
+the registry.  This allows the consumer to decode the message.
+
+| ![image]({{'/assets/pages/images/schema_validation_schema_registry.png' | absolute_url}}){:width="100%"} |
+|:-----------------------------------------------------------------:|
+|    *Kafka applications making use of a Schema Registry*     |
+
+
+#### So, what's the problem?
+
+If schemas are used correctly across the entire producing and consuming application estate, the potential for a
+poison message to be introduced into the system is greatly diminished.  However, the issue is that there's significant
+responsibility on the producing applications to configure and use schemas correctly.  There's nothing in the system to
+stop, say, a misconfigured producing application using the wrong schema or failing to use schemas at all.
+
+In Kafka, **brokers** have no knowledge of format of the messages (to the broker, a message is just opaque bytes) and they
+have no knowledge of the schema.  The brokers cannot help us enforce that producing applications use schemas and use them
+correctly.
+
+#### Kroxylicious Record Validation
+
+The Kroxylicious Record Validation Filter provides a solution to the problem. 
+
+The filter intercepts the produce requests sent from producing applications and subjects them to validation. If
+validation fail, the product request is rejected.  The producing application receives an error response.  The broker
+does not receive the rejected records.  In this way, one can organize that a poison message never enters the system,
+even if producing applications are misconfigured.
+
+The filter currently supports two types of validation:
+
+* _Schema validation_[^1] validates the content of the record against a schema.  
+* _SyntacticallyCorrectJson_ validation ensures the producer is producing messages that contain syntactically valid
```

Comment:
Maybe we shouldn't mention `SyntacticallyCorrectJson` in the use-case documentation at all.   I think it was really a POC that helped us prove the idea.

Reply:
Its one of those things where it helps the unstructured JSON crowd. So if we are taking a binary view and pushing schemas everywhere then I don't think there is any value in mentioning it here.  I'd be quite happy with something like 
> The filter currently supports two modes of operation
> 1. _Schema validation_[^1] validates the content of the record against a schema.   - For topics which have an entry in the centralised schema registry
> 2. _SyntacticallyCorrectJson_ validation ensures the producer is producing messages that contain syntactically valid JSON - For topics which do not have registered schemas.  

Having mentioned the 2nd option I don't think we need to go much further.